linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -1198,17 +1252,50 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
 
     fb.block() { objIsNullLabel =>
       primType match {
+        // For byte and short, use br_on_cast_fail with i31 then check the value
+        case ByteType | ShortType =>
+          val intValueLocal = fb.addLocal("intValue", Int32)
+
+          fb.block(RefType.anyref) { castFailLabel =>
+            fb += LocalGet(objParam)
+            fb += BrOnNull(objIsNullLabel)
+            fb += BrOnCastFail(castFailLabel, RefType.any, RefType.i31)
+
+            // Extract the i31 value
+            fb += I31GetS
+            fb += LocalTee(intValueLocal)
+
+            // if it sign-extends to itself
+            fb += LocalGet(intValueLocal)
+            if (primType == ByteType)
+              fb += I32Extend8S
+            else
+              fb += I32Extend16S
+            fb += I32Eq
+            fb.ifThen() {
+              // then success
+              if (isUnbox)
+                fb += LocalGet(intValueLocal)
+              else
+                fb += LocalGet(objParam)
+              fb += Return
+            }
+
+            // Fall through for CCE
```

Comment:
Hum, I think this comment is misleading, no? If someone writes `1` in JS and passes it to WASM, it will not be a i31 but it is a valid Byte?

That being said, do we have tests for this? :)

Reply:
Keeping open :) Do we have tests for this?

Reply:
We have plenty of tests that call native JS APIs returning `Byte`s and `Short`s, yes. They would throw under fatal/compliant semantics if the above code did not correctly identify JS `number`s that fit in `Byte`s and `Short`s.

One example I found:
https://github.com/scala-js/scala-js/blob/b79ebf80acbefb984f042b2031667ef82e297c46/test-suite/js/src/test/scala/org/scalajs/testsuite/typedarray/DataViewTest.scala#L54-L65
for `Byte`s, and the corresponding for `Short`s:
https://github.com/scala-js/scala-js/blob/b79ebf80acbefb984f042b2031667ef82e297c46/test-suite/js/src/test/scala/org/scalajs/testsuite/typedarray/DataViewTest.scala#L78-L90